systems/plants/setupLCPmex.cpp

```diff
@@ -0,0 +1,275 @@
+  }
+}
+
+//counts number of inclusions
+inline int getNumTrue(vector<bool> const & bools)
+{
+  int count = 0;
+  const int n = bools.size();
+  for (int x = 0; x < n; x++) {
+    if (bools[x]) {
+      count++;
+    }
+  }
+  return count;
+}
+
+//splits a vector into two based on inclusion mapping
+inline void partitionVector(vector<bool> const &indices, VectorXd const & v, VectorXd & included, VectorXd & excluded)
+{
+  const int n = indices.size();
+  const int count = getNumTrue(indices);
+
+  included = VectorXd::Zero(count);
+  excluded = VectorXd::Zero(n-count);
+  
+  int inclusionIndex = 0;
+  int exclusionIndex = 0;
+
+  for (int x = 0; x < n; x++) {
+    if (indices[x]) {
+      included[inclusionIndex++] = v[x];
+    } else {
+      excluded[exclusionIndex++] = v[x];
+    }
+  }
+}
+
+//splits a matrix into two based on a row inclusion mapping
+inline void partitionMatrix(vector<bool> const &indices, MatrixXd const & M, MatrixXd & included, MatrixXd & excluded)
+{   
+  const int n = indices.size();
+  const int count = getNumTrue(indices);
+  const int cols = M.cols();
+
+  included = MatrixXd::Zero(count, cols);
+  excluded = MatrixXd::Zero(n-count, cols);
+
+  int inclusionIndex = 0;
+  int exclusionIndex = 0;
+
+  for (int x = 0; x < n; x++) {
+    if (indices[x]) {
+      included.row(inclusionIndex++) = M.row(x);
+    } else {
+      excluded.row(exclusionIndex++) = M.row(x);
+    }
+  }
+}
+
+//builds a filtered matrix containing only the rows specified by indices
+inline void filterByIndices(vector<int> const &indices, MatrixXd const & M, MatrixXd & filtered)
+{
+  const int n = indices.size();
+  filtered = MatrixXd::Zero(n, M.cols());
+  for (int x = 0; x < n; x++) {
+    filtered.row(x) = M.row(indices[x]);
+  }
+}
+
+//[M, w, Mqdn, wqdn] = setupLCPmex(mex_model_ptr, q, qd, u, phiC, n, D, h, z_inactive_guess_tol)
+void mexFunction( int nlhs, mxArray *plhs[],int nrhs, const mxArray *prhs[] ) { 
+  RigidBodyManipulator *model= (RigidBodyManipulator*) getDrakeMexPointer(prhs[0]);
+  const int nq = model->num_positions;
+  const int nv = model->num_velocities;
+  const int numContactPairs = mxGetNumberOfElements(prhs[4]);
+  const int mC = mxGetNumberOfElements(prhs[6]);
+  const double z_inactive_guess_tol = static_cast<double>(mxGetScalar(prhs[8]));
+  const double h = static_cast<double>(mxGetScalar(prhs[7]));
+  const Map<VectorXd> q(mxGetPr(prhs[1]),nq);
+  const Map<VectorXd> v(mxGetPr(prhs[2]),nv);
+  const Map<VectorXd> u(mxGetPr(prhs[3]),model->B.cols());
+  
+  const Map<VectorXd> phiC(mxGetPr(prhs[4]),numContactPairs);
+  const Map<MatrixXd> n(mxGetPr(prhs[5]), numContactPairs, nq);
+
+  VectorXd C, phiL, phiP, phiL_possible, phiC_possible, phiL_check, phiC_check;
+  MatrixXd H, B, JP, JL, JL_possible, n_possible, JL_check, n_check;
+
+  manipulatorDynamics(model, q, v, H, C, B);
+
+  model->positionConstraints(phiP, JP);
+  model->jointLimitConstraints(q, phiL, JL);
+  MatrixXd Hinv = H.inverse();
```

Comment:
I think using a solve may be faster, even if you're making use of Hinv in a number of place. Should probably use a decomposition that exploits the fact that H is positive definite, such as Cholesky (LLT): 
http://eigen.tuxfamily.org/dox/classEigen_1_1LLT.html